File: core/src/main/java/com/terralite/core/registry/ResourceKey.java
package com.terralite.core.registry;

import java.util.Objects;

public record ResourceKey<T>(RegistryKey<T> registry, ResourceId id) {
    public ResourceKey {
        Objects.requireNonNull(registry, "registry");
        Objects.requireNonNull(id, "id");
    }

    public static <T> ResourceKey<T> of(RegistryKey<T> registry, ResourceId id) {
        return new ResourceKey<>(registry, id);
    }

    public static <T> ResourceKey<T> of(RegistryKey<T> registry, String id) {
        return of(registry, ResourceId.parse(id));
    }
}

File: core/src/commonMain/kotlin/com/terralite/core/registry/ResourceKey.kt
package com.terralite.core.registry

/** A fully-qualified key that combines the target [registry] with a specific entry [id]. */
@JvmRecord
data class ResourceKey<T>(val registry: RegistryKey<T>, val id: ResourceId) {

    companion object {
        @JvmStatic
        fun <T> of(registry: RegistryKey<T>, id: ResourceId): ResourceKey<T> = ResourceKey(registry, id)

        @JvmStatic
        fun <T> of(registry: RegistryKey<T>, id: String): ResourceKey<T> = ResourceKey(registry, ResourceId.parse(id))
    }
}
